File: app/src/main/kotlin/tech/mmarca/openvitals/data/repository/contract/HydrationRepository.kt
package tech.mmarca.openvitals.data.repository.contract

import java.time.LocalDate
import tech.mmarca.openvitals.core.period.PeriodLoadQuery
import tech.mmarca.openvitals.domain.model.BeverageCategory
import tech.mmarca.openvitals.domain.model.CustomHydrationDrink
import tech.mmarca.openvitals.domain.model.DailyHydration
import tech.mmarca.openvitals.domain.model.HydrationEntry
import tech.mmarca.openvitals.domain.model.HydrationWriteRequest
import tech.mmarca.openvitals.domain.model.RefreshMode
import tech.mmarca.openvitals.domain.query.HydrationPeriodData

interface HydrationRepository {
    val hydrationWritePermissions: Set<String>

    fun hydrationContainerVolumeMilliliters(): Map<String, Double>

    fun setHydrationContainerVolumeMilliliters(containerId: String, milliliters: Double)

    fun lastCustomHydrationAmountMilliliters(): Double?

    fun setLastCustomHydrationAmountMilliliters(milliliters: Double)

    /** The last cup sizes logged, most recent first, for one-tap quick adds. */
    fun recentHydrationAmountsMilliliters(): List<Double>

    fun recordRecentHydrationAmountMilliliters(milliliters: Double)

    fun customHydrationDrinks(): List<CustomHydrationDrink>

    fun saveCustomHydrationDrink(drink: CustomHydrationDrink)

    fun deleteCustomHydrationDrink(drinkId: String)

    fun reorderCustomHydrationDrinks(drinkIds: List<String>)

    fun moveCustomHydrationDrinkToCategory(drinkId: String, category: BeverageCategory?)

    fun hydrationDailyGoalLiters(): Double

    suspend fun loadHydrationPeriod(
        query: PeriodLoadQuery,
        refreshMode: RefreshMode = RefreshMode.NORMAL,
    ): HydrationPeriodData

    suspend fun loadDailyHydration(start: LocalDate, end: LocalDate): List<DailyHydration>

    suspend fun loadHydrationEntries(start: LocalDate, end: LocalDate): List<HydrationEntry>

    suspend fun hasHydrationWritePermission(): Boolean

    suspend fun writeHydrationEntry(request: HydrationWriteRequest): String

    suspend fun loadHydrationEntry(id: String): HydrationEntry?

    suspend fun updateHydrationEntry(id: String, request: HydrationWriteRequest)

    suspend fun deleteHydrationEntry(id: String)

    /** Rolls back a just-written record when the paired nutrition write fails. */
    suspend fun deleteHydrationEntryByClientRecordId(clientRecordId: String)
}

File: lib/data/repository/contract/hydration_repository.dart
import '../../../core/period/period_load_query.dart';
import '../../../core/result/result.dart';
import '../../../core/time/local_date.dart';
import '../../../domain/model/caffeine_models.dart';
import '../../../domain/model/nutrition_models.dart';
import '../../../domain/model/refresh_mode.dart';
import '../../../domain/query/hydration_period_data.dart';

/// Port of the Kotlin `HydrationRepository` contract.
///
/// Fallible operations return [Result]; the synchronous probes (the write
/// permission set, the preference-backed container/goal reads and writes)
/// read cached state and cannot fail, so they stay bare.
abstract interface class HydrationRepository {
  Set<String> get hydrationWritePermissions;

  Map<String, double> hydrationContainerVolumeMilliliters();

  void setHydrationContainerVolumeMilliliters(
    String containerId,
    double milliliters,
  );

  double? lastCustomHydrationAmountMilliliters();

  void setLastCustomHydrationAmountMilliliters(double milliliters);

  /// The drink catalog: the seeded presets (still/gasified water plus the
  /// CaffeineHealth beverage list) merged with the user's own drinks.
  ///
  /// Asynchronous because the drift-backed `BeverageStore` is. The Kotlin
  /// `beverageStore.beverages()` is a blocking Room read, which is why its
  /// contract is synchronous.
  Future<Result<List<CustomHydrationDrink>>> customHydrationDrinks();

  Future<Result<void>> saveCustomHydrationDrink(CustomHydrationDrink drink);

  Future<Result<void>> deleteCustomHydrationDrink(String drinkId);

  Future<Result<void>> reorderCustomHydrationDrinks(List<String> drinkIds);

  Future<Result<void>> moveCustomHydrationDrinkToCategory(
    String drinkId,
    CaffeineSourceCategory? category,
  );

  double hydrationDailyGoalLiters();

  Future<Result<HydrationPeriodData>> loadHydrationPeriod(
    PeriodLoadQuery query, {
    RefreshMode refreshMode = RefreshMode.normal,
  });

  Future<Result<List<DailyHydration>>> loadDailyHydration(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<List<HydrationEntry>>> loadHydrationEntries(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<bool>> hasHydrationWritePermission();

  Future<Result<String>> writeHydrationEntry(HydrationWriteRequest request);

  Future<Result<HydrationEntry?>> loadHydrationEntry(String id);

  Future<Result<void>> updateHydrationEntry(
    String id,
    HydrationWriteRequest request,
  );

  Future<Result<void>> deleteHydrationEntry(String id);
}
